### PR TITLE
AP: Deprecate external refs

### PR DESCRIPTION
**Pull request by rafwegv**
### Proposed changes
Add a warning when an external refs field is detected in the policy.
This is the first stage of deprecation. In next release the policy will become invalid , and further down the line the fields will be removed from the crd.

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto master
- [x] I will ensure my PR is targeting the master branch and pulling from my branch from my own fork

### CODE REVIEW

**Comment by rafwegv** [comment]
Hi @soneillf5 @pleshakov I would appreciate input on where would be a good place to document this change, to make any customers aware of the deprecation.

**Comment by brianehlert** [comment]
> 

Bare minimum is release notes.
I would add a deprecation note into the docs where the configuration option is presented to make it very clear.

**Comment by AmitShnek** [comment]
@rafwegv ?

**Comment by rafwegv** [comment]
Hi @brianehlert i have added a note in the config docs. As I understand that will be then referenced in the release notes.

**Comment by soneillf5** [comment]
@rafwegv Are there any tests associated with this change? How do we know this works and doesn't break in the future?

**Comment by rafwegv** [comment]
@soneillf5 This feature does not touch the kic code, it goes straight to the file on storage without any checks. This PR is only meant to warn the admin that this feature is going to go away in the future.

File: docs/content/app-protect/configuration.md [hunk 1]
```diff
@@ -29,6 +29,8 @@ You can define App Protect policies for your Ingress resources by creating an `A
  > **Note**: [The Advanced gRPC Protection for Unary Traffic](/nginx-app-protect/configuration/#advanced-grpc-protection-for-unary-traffic) only supports providing an `idl-file` inline. The fields `policy.idl-files[].link`, `policy.idl-files[].$ref`, and 
  `policy.idl-files[].file` are not supported. The IDL file should be provided in field `policy.idl-files[].contents`. The value of this field can be base64 encoded. In this case the field `policy.idl-files[].isBase64` should be set to `true`.
 
+ > **Note**: [External References](/nginx-app-protect/configuration-guide/configuration/#external-references) are deprecated and will may not be supported in future releases.
+
```
[the remark this answers is not in the document]
Comment (rafwegv) [reply docs/content/app-protect/configuration.md:33]: @ciarams87 fixed

File: internal/k8s/appprotect/app_protect_resources.go [hunk 2]
```diff
@@ -148,3 +188,18 @@ func validateAppProtectUserSig(userSig *unstructured.Unstructured) error {
 func GetNsName(obj *unstructured.Unstructured) string {
 	return obj.GetNamespace() + "/" + obj.GetName()
 }
+
+func checkForExtRefs(policy *unstructured.Unstructured) ([]string, error) {
```
Comment (lucacome) [inline internal/k8s/appprotect/app_protect_resources.go:192]: Hi @rafwegv I think it makes sense to extend the unit tests to test this new function, what do you think?

**Review by lucacome — CHANGES_REQUESTED** [review]
Is this PR still relevant? I couldn't find any reference to the deprecation of external reference in NAP WAF docs. If so I left a comment.

File: internal/k8s/appprotect/app_protect_resources.go (hunk 2 above)
Comment (rafwegv) [reply internal/k8s/appprotect/app_protect_resources.go:192]: Hi @lucacome the deprecation is only true for KIC. I corrected the doc to reflect that. also added unit test as sugested.

File: docs/content/app-protect/configuration.md [hunk 3]
```diff
@@ -30,7 +30,9 @@ You can define App Protect policies for your Ingress resources by creating an `A
  > **Note**: [The Advanced gRPC Protection for Unary Traffic](/nginx-app-protect/configuration/#advanced-grpc-protection-for-unary-traffic) only supports providing an `idl-file` inline. The fields `policy.idl-files[].link`, `policy.idl-files[].$ref`, and
  `policy.idl-files[].file` are not supported. The IDL file should be provided in field `policy.idl-files[].contents`. The value of this field can be base64 encoded. In this case the field `policy.idl-files[].isBase64` should be set to `true`.
 
-To add any [App Protect policy](/nginx-app-protect/declarative-policy/policy/) to an Ingress resource:
+ > **Note**: [External References](/nginx-app-protect/configuration-guide/configuration/#external-references) in KIC are deprecated and will not be supported in future releases.
```
Comment (lucacome) [inline docs/content/app-protect/configuration.md:33]:
```suggestion
 > **Note**: [External References](/nginx-app-protect/configuration-guide/configuration/#external-references) in the Ingress Controller are deprecated and will not be supported in future releases.
```

**Review by lucacome — APPROVED** [review]

**Comment by lucacome** [comment]
@rafwegv I noticed that the are a lot of changes to unrelated files, but it looks like it's about formatting and spaces so I approved it anyway

**Review by ciarams87 — APPROVED** [review]

**Pull request merged**